### ISSUE TITLE
what is your fitness function?

### ISSUE THREAD

**Issue by dashujuzha** [tytu16/Kaggle-Fish-Detection#1]
How to  understand    this code :"fitness = .7*(loss_fit) + .3*(time_cost)"

**Comment by t2techno** [comment]
My idea was to use a weighted combination of the loss from training and the time taken to run one training epoch as my final fitness score for a given network. I wanted the training_loss to have a larger effect on the final fitness score, so I scaled them with a heavier weight on the training loss.
In hindsight, I imagine it would have the same effect to simply scale the time_cost down, and leave my training loss as is.

**Comment by dashujuzha** [comment]
Thanks very much

**Comment by dashujuzha** [comment]
How do you initialize the population, why aren't the Numbers fixed?



Builds a chromosome from random
The first gene determines layer types:
    0: Inception module
    1: Convolution layer
    2: Fully Connected(only available for the last layer)
    
    Inception = [layer_type,depth_1x1,filter_depth1,dropout]
        layer_type:   0
        depth_1x1:    2^n, 3 <= n <= 6
        filt_power_1: 2^m, 4 <= m <= 8
        dropout:     True or False
        
    Convolution = [layer_type,filter_size,num_filters,filter_stride,pool_size,dropout]
        layer_type:   1
        filter_size:  2*n+1, 2 <= n <= 4
        num_filters:  2^m,   3 <= m <= 8
        filter_stride:j,     1 <= j <= n
        pool_size:    k,     0 <= k <= 3, does not use pooling if k < 2
        dropout:      True or False
        
    Fully Connected = [layer_type,layer_size,dropout]
        layer_type: 2
        layer_size: 6 <= n <= 9, 2^n
        dropout:    True or False
"""

**Comment by t2techno** [comment]
The numbers aren't fixed because I was curious how the hyper-parameters would compare to things I had come up with in my hand-tuned networks. I did impose some semblance of order, such as making the filter size odd(so there would be a center pixel in the square) and most other parameters powers of two. 
I could have picked the hyper-parameters for each layer beforehand, and simply randomized which layers it used, but when creating networks by hand, I usually don't use the same hyper-parameters for each layer. I figured if they were static, the overall potential to learn could be stunted.

The population is initialized in the run_world function:

```python
        # generate whole population from random if first generation
        if gen_num == 1:
            print("Generating initial population...")
            for i in range(pop_size):
                population.append(create_new_chromosome(chr_size=5))
            print_pop(population)

        # otherwise, create new population via mating of top 3 individuals
        # assuming pre-sorted based on fitness
        else:
            print("Mating top 3...")
            #randomly generated mutation_rate, from between 0 and .7
            population = mate(population[:3],mutation_rate=random.random(0,.7))
            print_pop(population)
```

The first generation is created entirely at random, and thereafter is modified via mutations and cross-over between two parent networks. It would probably be beneficial for me to go in and add the option to start from a list of already created networks, that way I could continue using a population that has been previously generated.

**Comment by t2techno** [comment]
As I go through this program again, another thing I notice that should be adjusted is that I never compensate for when dropout is used. Given two identical networks trained for the same amount of time with and without dropout, the one with dropout should score worse since it was attempting to use less information. I should add extra training-epochs in proportion to the number of layers with dropout.
